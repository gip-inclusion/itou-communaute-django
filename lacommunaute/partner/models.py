from django.conf import settings
from django.db import models
from django.urls import reverse
from django.utils.encoding import force_str
from django.utils.text import slugify
from django.utils.translation import gettext_lazy as _
from machina.models.abstract_models import DatedModel
from machina.models.fields import MarkupTextField
from storages.backends.s3boto3 import S3Boto3Storage

from lacommunaute.utils.validators import validate_image_size


class Partner(DatedModel):
    name = models.CharField(max_length=100, verbose_name="Nom", unique=True)
    slug = models.SlugField(max_length=255, verbose_name="Slug")
    short_description = models.CharField(
        max_length=400, blank=True, null=True, verbose_name="Description courte (SEO)"
    )
    description = MarkupTextField(verbose_name="Description", null=True, blank=True)
    logo = models.ImageField(
        storage=S3Boto3Storage(bucket_name=settings.AWS_STORAGE_BUCKET_NAME, file_overwrite=False),
        validators=[validate_image_size],
        upload_to="logos/",
        help_text="1200x600 recommand√©",
    )
    url = models.URLField(verbose_name="Lien vers le site du partenaire", null=True, blank=True)

    objects = models.Manager()

    class Meta:
        verbose_name = _("Partner")
        verbose_name_plural = _("Partners")
        ordering = ["-updated"]

    def __str__(self):
        return f"{self.name}"

    def save(self, *args, **kwargs):
        self.slug = slugify(force_str(self.name), allow_unicode=True)
        super().save(*args, **kwargs)

    def get_absolute_url(self):
        return reverse("partner:detail", kwargs={"slug": self.slug, "pk": self.pk})
