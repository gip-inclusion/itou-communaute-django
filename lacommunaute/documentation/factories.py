import factory
from faker import Faker

from lacommunaute.documentation.models import Category, Document, DocumentRating


faker = Faker()


class AbstractDocumentationFactory(factory.django.DjangoModelFactory):
    name = factory.Faker("name")
    description = factory.Faker("sentence", nb_words=100)
    short_description = factory.Faker("sentence", nb_words=10)
    image = factory.django.ImageField(filename="banner.jpg")


class CategoryFactory(AbstractDocumentationFactory):
    class Meta:
        model = Category

    class Params:
        for_snapshot = factory.Trait(
            name="Test Category", description="Test description", short_description="Test short description"
        )


class DocumentFactory(AbstractDocumentationFactory):
    category = factory.SubFactory(CategoryFactory)

    class Meta:
        model = Document

    class Params:
        for_snapshot = factory.Trait(
            name="Test Document",
            description="Test description",
            short_description="Test short description",
            category=factory.SubFactory(CategoryFactory, for_snapshot=True),
        )

    @factory.post_generation
    def with_tags(self, create, extracted, **kwargs):
        if not create or not extracted:
            return

        if isinstance(extracted, list):
            for tag in extracted:
                self.tags.add(tag)


class DocumentRatingFactory(factory.django.DjangoModelFactory):
    document = factory.SubFactory(DocumentFactory)
    rating = factory.Faker("random_int", min=1, max=5)

    class Meta:
        model = DocumentRating
        skip_postgeneration_save = True

    @factory.post_generation
    def set_created(self, create, extracted, **kwargs):
        if extracted:
            self.created = extracted
            self.save()
