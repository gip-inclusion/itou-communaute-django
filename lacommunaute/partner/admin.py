from django.contrib import admin

from lacommunaute.partner.models import Partner


@admin.register(Partner)
class PartnerAdmin(admin.ModelAdmin):
    list_display = ("name", "url", "created")
    search_fields = ("name",)
    list_filter = ("created",)
