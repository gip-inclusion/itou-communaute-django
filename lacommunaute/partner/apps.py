from django.apps import AppConfig


class PartenerAppConfig(AppConfig):
    label = "partner"
    name = "lacommunaute.partner"
    verbose_name = "Partner"
    verbose_name_plural = "Partners"
