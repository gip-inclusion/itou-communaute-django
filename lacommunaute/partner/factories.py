import factory
from faker import Faker

from lacommunaute.partner.models import Partner


faker = Faker()


class PartnerFactory(factory.django.DjangoModelFactory):
    class Meta:
        model = Partner

    name = factory.LazyAttribute(lambda x: faker.company())
    short_description = factory.LazyAttribute(lambda x: faker.text(max_nb_chars=200))
    description = factory.LazyAttribute(
        lambda x: "\n".join([f"### part {i} \n{faker.paragraph(nb_sentences=20)}" for i in range(2)])
    )
    url = factory.LazyAttribute(lambda x: faker.url())

    class Params:
        with_image = factory.Trait(image=factory.django.ImageField(filename="test.jpg"))
        for_snapshot = factory.Trait(
            name="Best Partner Ever",
            description="### h3 long MD description \n lorem ipsum dolor sit amet, consectetur adipiscing elit.",
            short_description="short description for SEO",
            url="https://www.best-partner-ever.com",
        )
