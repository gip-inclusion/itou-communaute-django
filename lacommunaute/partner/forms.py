from django import forms
from django.conf import settings

from lacommunaute.partner.models import Partner


class PartnerForm(forms.ModelForm):
    logo = forms.ImageField(
        required=False,
        label="Logo, format 200 x 200 pixels recommand√©",
        widget=forms.FileInput(attrs={"accept": settings.SUPPORTED_IMAGE_FILE_TYPES.keys()}),
    )

    class Meta:
        model = Partner
        fields = ("name", "short_description", "description", "logo", "url")
