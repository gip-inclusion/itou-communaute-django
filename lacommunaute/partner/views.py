from django.views.generic import DetailView, ListView

from lacommunaute.forum.models import Forum
from lacommunaute.partner.models import Partner
from lacommunaute.utils.perms import forum_visibility_content_tree_from_forums


class PartnerListView(ListView):
    model = Partner
    template_name = "partner/list.html"
    context_object_name = "partners"
    paginate_by = 8 * 3


class PartnerDetailView(DetailView):
    model = Partner
    template_name = "partner/detail.html"
    context_object_name = "partner"

    def get_context_data(self, **kwargs):
        context = super().get_context_data(**kwargs)
        context["sub_forums"] = forum_visibility_content_tree_from_forums(
            self.request, Forum.objects.filter(partner=self.object)
        )
        return context
