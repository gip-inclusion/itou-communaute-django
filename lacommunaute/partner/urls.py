from django.urls import path

from lacommunaute.partner.views import PartnerDetailView, PartnerListView


app_name = "partner"


urlpatterns = [
    path("", PartnerListView.as_view(), name="list"),
    path("<str:slug>-<int:pk>/", PartnerDetailView.as_view(), name="detail"),
]
