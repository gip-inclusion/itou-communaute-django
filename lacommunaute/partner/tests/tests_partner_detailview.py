import pytest  # noqa

from lacommunaute.forum.factories import ForumFactory
from lacommunaute.partner.factories import PartnerFactory
from lacommunaute.users.factories import UserFactory
from lacommunaute.utils.testing import parse_response_to_soup


@pytest.mark.parametrize(
    "user,snapshot_name",
    [
        (lambda: None, "partner_detailview as anonymous"),
        (lambda: UserFactory(), "partner_detailview"),
        (lambda: UserFactory(is_superuser=True), "partner_detailview as superuser"),
    ],
)
def test_partner_detailview(client, db, snapshot, user, snapshot_name):
    partner = PartnerFactory(for_snapshot=True, with_image=True)
    user = user()
    if user:
        client.force_login(user)
    response = client.get(partner.get_absolute_url())
    assert response.status_code == 200
    assert str(
        parse_response_to_soup(response, selector="main", replace_img_src=True, replace_in_href=[partner])
    ) == snapshot(name=snapshot_name)


def test_partner_with_forums(client, db, snapshot):
    partner = PartnerFactory(for_snapshot=True, with_image=True)
    forum = ForumFactory(with_partner=partner, with_public_perms=True, with_tags=["sun", "moon"], for_snapshot=True)
    response = client.get(partner.get_absolute_url())
    assert response.status_code == 200
    assert str(
        parse_response_to_soup(response, selector="main", replace_img_src=True, replace_in_href=[forum])
    ) == snapshot(name="partner_detailview_with_forums")
