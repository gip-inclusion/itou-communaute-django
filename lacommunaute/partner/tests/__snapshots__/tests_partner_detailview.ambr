# serializer version: 1
# name: test_partner_detailview[<lambda>-partner_detailview as anonymous][partner_detailview as anonymous]
  '''
  <main class="s-main" id="main" role="main">
              
                  
              
              
      
  <div class="container">
      <nav aria-label="Fil d'ariane" class="c-breadcrumb">
          <ol class="breadcrumb">
              <li class="breadcrumb-item">Retourner vers</li>
              <li class="breadcrumb-item">
                  <a class="matomo-event" data-matomo-action="view" data-matomo-category="engagement" data-matomo-option="partner-list" href="/partenaires/">Nos partenaires</a>
              </li>
          </ol>
      </nav>
  </div>
  
  
              
                  
      <section class="s-title-01 mt-lg-5">
          <div class="s-title-01__container container">
              <div class="s-title-01__row row">
                  <div class="s-title-01__col col-12">
                      <div class="d-flex gap-3 align-items-center mb-2">
                          
                              <img alt="Best Partner Ever" class="img-fluid" src="[img src]" width="200"/>
                          
                          <h1 class="mb-0">Best Partner Ever</h1>
                      </div>
                      
                      <h2 class="mt-3">short description for SEO</h2>
                  </div>
              </div>
          </div>
      </section>
      <section class="s-section">
          <div class="s-section__container container">
              <div class="s-section__row row">
                  <div class="col-12">
                      <article class="textarea_cms_md"><h3>h3 long MD description</h3>
  
  <p>lorem ipsum dolor sit amet, consectetur adipiscing elit.</p></article>
                      
                          <div class="mt-3">
                              <a class="has-external-link" href="https://www.best-partner-ever.com" rel="noopener noreferrer" target="_blank">https://www.best-partner-ever.com</a>
                          </div>
                      
                  </div>
              </div>
              
          </div>
      </section>
  
              
          </main>
  '''
# ---
# name: test_partner_detailview[<lambda>-partner_detailview as superuser][partner_detailview as superuser]
  '''
  <main class="s-main" id="main" role="main">
              
                  
              
              
      
  <div class="container">
      <nav aria-label="Fil d'ariane" class="c-breadcrumb">
          <ol class="breadcrumb">
              <li class="breadcrumb-item">Retourner vers</li>
              <li class="breadcrumb-item">
                  <a class="matomo-event" data-matomo-action="view" data-matomo-category="engagement" data-matomo-option="partner-list" href="/partenaires/">Nos partenaires</a>
              </li>
          </ol>
      </nav>
  </div>
  
  
              
                  
      <section class="s-title-01 mt-lg-5">
          <div class="s-title-01__container container">
              <div class="s-title-01__row row">
                  <div class="s-title-01__col col-12">
                      <div class="d-flex gap-3 align-items-center mb-2">
                          
                              <img alt="Best Partner Ever" class="img-fluid" src="[img src]" width="200"/>
                          
                          <h1 class="mb-0">Best Partner Ever</h1>
                      </div>
                      
                          <a href="/admin/partner/partner/[PK of Partner]/change/"><small>Mettre à jour</small></a>
                      
                      <h2 class="mt-3">short description for SEO</h2>
                  </div>
              </div>
          </div>
      </section>
      <section class="s-section">
          <div class="s-section__container container">
              <div class="s-section__row row">
                  <div class="col-12">
                      <article class="textarea_cms_md"><h3>h3 long MD description</h3>
  
  <p>lorem ipsum dolor sit amet, consectetur adipiscing elit.</p></article>
                      
                          <div class="mt-3">
                              <a class="has-external-link" href="https://www.best-partner-ever.com" rel="noopener noreferrer" target="_blank">https://www.best-partner-ever.com</a>
                          </div>
                      
                  </div>
              </div>
              
          </div>
      </section>
  
              
          </main>
  '''
# ---
# name: test_partner_detailview[<lambda>-partner_detailview][partner_detailview]
  '''
  <main class="s-main" id="main" role="main">
              
                  
              
              
      
  <div class="container">
      <nav aria-label="Fil d'ariane" class="c-breadcrumb">
          <ol class="breadcrumb">
              <li class="breadcrumb-item">Retourner vers</li>
              <li class="breadcrumb-item">
                  <a class="matomo-event" data-matomo-action="view" data-matomo-category="engagement" data-matomo-option="partner-list" href="/partenaires/">Nos partenaires</a>
              </li>
          </ol>
      </nav>
  </div>
  
  
              
                  
      <section class="s-title-01 mt-lg-5">
          <div class="s-title-01__container container">
              <div class="s-title-01__row row">
                  <div class="s-title-01__col col-12">
                      <div class="d-flex gap-3 align-items-center mb-2">
                          
                              <img alt="Best Partner Ever" class="img-fluid" src="[img src]" width="200"/>
                          
                          <h1 class="mb-0">Best Partner Ever</h1>
                      </div>
                      
                      <h2 class="mt-3">short description for SEO</h2>
                  </div>
              </div>
          </div>
      </section>
      <section class="s-section">
          <div class="s-section__container container">
              <div class="s-section__row row">
                  <div class="col-12">
                      <article class="textarea_cms_md"><h3>h3 long MD description</h3>
  
  <p>lorem ipsum dolor sit amet, consectetur adipiscing elit.</p></article>
                      
                          <div class="mt-3">
                              <a class="has-external-link" href="https://www.best-partner-ever.com" rel="noopener noreferrer" target="_blank">https://www.best-partner-ever.com</a>
                          </div>
                      
                  </div>
              </div>
              
          </div>
      </section>
  
              
          </main>
  '''
# ---
# name: test_partner_with_forums[partner_detailview_with_forums]
  '''
  <main class="s-main" id="main" role="main">
              
                  
              
              
      
  <div class="container">
      <nav aria-label="Fil d'ariane" class="c-breadcrumb">
          <ol class="breadcrumb">
              <li class="breadcrumb-item">Retourner vers</li>
              <li class="breadcrumb-item">
                  <a class="matomo-event" data-matomo-action="view" data-matomo-category="engagement" data-matomo-option="partner-list" href="/partenaires/">Nos partenaires</a>
              </li>
          </ol>
      </nav>
  </div>
  
  
              
                  
      <section class="s-title-01 mt-lg-5">
          <div class="s-title-01__container container">
              <div class="s-title-01__row row">
                  <div class="s-title-01__col col-12">
                      <div class="d-flex gap-3 align-items-center mb-2">
                          
                              <img alt="Best Partner Ever" class="img-fluid" src="[img src]" width="200"/>
                          
                          <h1 class="mb-0">Best Partner Ever</h1>
                      </div>
                      
                      <h2 class="mt-3">short description for SEO</h2>
                  </div>
              </div>
          </div>
      </section>
      <section class="s-section">
          <div class="s-section__container container">
              <div class="s-section__row row">
                  <div class="col-12">
                      <article class="textarea_cms_md"><h3>h3 long MD description</h3>
  
  <p>lorem ipsum dolor sit amet, consectetur adipiscing elit.</p></article>
                      
                          <div class="mt-3">
                              <a class="has-external-link" href="https://www.best-partner-ever.com" rel="noopener noreferrer" target="_blank">https://www.best-partner-ever.com</a>
                          </div>
                      
                  </div>
              </div>
              
                  <div class="s-section__row row mt-5">
                      <div class="col-12">
                          <h2>Leurs contributions dans la communauté</h2>
                      </div>
                      <div class="col-12"><div class="container-fluid" id="subforums_area">
      
      <div class="row mt-4" id="documentation-category-subforums">
          
              <div class="col-12 col-md-4 mb-3 mb-md-5">
                  <div class="card c-card has-one-link-inside h-100">
                      
                      <div class="card-body pb-0">
                          <p class="h3 lh-base">Test Forum</p>
                          <div>
      
          <span class="tag bg-info-lighter text-info">sun</span>
      
  
      
          <span class="tag bg-info-lighter text-info">moon</span>
      
  
  </div>
                          <div class="mt-3">Test description</div>
                      </div>
                      <div class="card-footer text-end">
                          <a class="btn btn-sm btn-ico btn-link stretched-link matomo-event" data-matomo-action="view" data-matomo-category="engagement" data-matomo-option="fiches_techniques" href="/forum/test-forum-[PK of Forum]/">
                              <i class="ri-arrow-right-line ri-lg"></i>
                          </a>
                      </div>
                  </div>
              </div>
          
      </div>
  </div>
  </div>
                  </div>
              
          </div>
      </section>
  
              
          </main>
  '''
# ---
